File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        self.res = 0
        self.path = [str(root.val)]
        
        def dfs(node):
            
            if not node.left and not node.right:
                self.res += int("".join(self.path))
                return
            
            for child in node.left, node.right:
                if child:
                    self.path.append(str(child.val))
                    dfs(child)
                    self.path.pop()                              
        
        dfs(root)
        return self.res
    

File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer}
def sum_numbers(root)
    $res = 0
    
    def dfs(node, cur_sum)
        if not node.left and not node.right
            cur_sum = cur_sum * 10 + node.val
            $res += cur_sum
        end
        
        if node.left
            dfs(node.left, cur_sum * 10 + node.val)
        end
        
        if node.right
            dfs(node.right, cur_sum * 10 + node.val)
        end    
    end
    
    dfs(root, 0)
    $res
    
end